OAuth: per-service expiration in expires_in

gonzalad opened the pull request:
From OAuth specifications, expires_in from token
and authorization endpoints must be equal to
accessToken expiration.

Before this fix, expires_in was always set to
global oauth accessToken expiration and not
per-service expiration.

gonzalad commented:
Previous commit resolved source conflicts with master branch

gonzalad commented:
The stale bot automatically closed this pull request. 
AFAIK, the issue reported here is not solved:

`
From OAuth specifications, expires_in from token
and authorization endpoints must be equal to
accessToken expiration
`

Can someone either:

- reopen this issue (I don't find a way to do it myself)
- or let me know if I need to rework this PR in order for it to be accepted
- or otherwise point me to an issue tracker for Cas to track this issue 

Thanks,
Adrian

mmoayyed commented:
Hi there,

Thanks for the follow-up. Please review the following:

- Make sure CI passes, and that you have tests for all changes.
- Make sure your test coverage stays above 80%
- Please describe an overview of the changes you have made with a brief explanation of how it all ties together.
- For reference provide a link to the OAuth spec that outlines the requirement.

Thanks again!

gonzalad commented:
First all sorry for my last comment: I was mistaken about the issue this ticket address.

The issue this ticket addresses is per-service access token expiration in expires_in.

ATM CAS token and authorization endpoints return the global AT expiration.

This PR just fixes that: it returns the per-service AT expiration or (if no per-service expiration is configured) 
the global AT expiration.

Now, I'll try to answer your questions:

> * Please describe an overview of the changes you have made with a brief explanation of how it all ties together.

**OAuth20AccessToken and OAuth20DefaultAccessToken**

  I added the getExpiresIn() method in OAuth20AccessToken (and OAuth20DefaultAccessToken) to get the expiresIn value in a central way since it will be used by multiple controllers.

  The implementation is
```
    @Override
    public long getExpiresIn() {
        return getExpirationPolicy().getTimeToLive();
    }
```
  the expirationPolicy was already set by OAuth20DefaultAccessTokenFactory and computed by in determineExpirationPolicyForService().

  If there's a per-service expiration policy, then it returns this per-service policy.
  If not, it returns the global policy.

  Everything was already done at that point, we only needed to expose the expiresIn value. 

**Controllers and their companion objects**

I updated all controllers (or their responseBuilder companion objects) to use accessToken.getExpiresIn() instead of the global AT expiration (atPolicy.buildTicketExpirationPolicy().getTimeToLive()).

  Sample classes: OAuth20AccessTokenEndpointController, OAuth20ResourceOwnerCredentialsResponseBuilder

  Note: in the case where not AT is returned, we don't return the expiresIn attribute anymore (see OAuth20DefaultAccessTokenResponseGenerator).

**Remaining modifications**

The remaining modifications were about removing the injection of ExpirationPolicyBuilder<OAuth20AccessToken> in different
  class since the expiration is now computed centrally in the OAuth20AccessToken.

Sample classes: OidcImplicitIdTokenAuthorizationResponseBuilder, OidcConfiguration



> * Make sure CI passes, and that you have tests for all changes.

I see the CI some issues https://travis-ci.org/github/apereo/cas/builds/684212092

There's a single error in OAuth category: https://travis-ci.org/github/apereo/cas/jobs/684212119

Looking at that I see an error in

```
> Task :support:cas-server-support-oauth:testOAuth
OAuth20AuthorizeEndpointControllerTests > verifyPerServiceExpiration() FAILED
    java.lang.ClassCastException: class org.pac4j.cas.profile.CasProfile cannot be cast to class java.util.Map (org.pac4j.cas.profile.CasProfile is in unnamed module of loader 'app'; java.util.Map is in module java.base of loader 'bootstrap')
        at org.pac4j.core.profile.ProfileManager.lambda$retrieveAll$1(ProfileManager.java:121)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at org.pac4j.core.profile.ProfileManager.retrieveAll(ProfileManager.java:121)
        at org.pac4j.core.profile.ProfileManager.get(ProfileManager.java:56)
        at org.apereo.cas.support.oauth.web.endpoints.OAuth20AuthorizeEndpointController.isRequestAuthenticated(OAuth20AuthorizeEndpointController.java:44)
        at org.apereo.cas.support.oauth.web.endpoints.OAuth20AuthorizeEndpointController.handleRequest(OAuth20AuthorizeEndpointController.java:62)
```

I don't see how it can be related to the current PR.
Seems some class from pac4j support module leaks here, but I don't see from where.
Don't know if there's much I can do :(


There are also errors in Simple Unit Test category: https://travis-ci.org/github/apereo/cas/jobs/684212112
but not related at all with this PR:

- StaticUserGraphicalAuthenticationRepositoryTests > verifyBadImage() 
- HazelcastHealthIndicatorTests
- CloudWatchAppenderSpecTests
...


> * Make sure your test coverage stays above 80%

Looking at codecov report, I don't see an impacted class with less than 80%
https://codecov.io/gh/apereo/cas/pull/4776/diff

If there's something I need to improve, point me please at the reports because I don't see what (I'm a bit lost with the huge CAS codebase)


Thanks,
Adrian

mmoayyed commented:
Outstanding. Thanks very much for the notes.

I'll take a look at tests tomorrow. Goal should be get CI passing on the master branch, and then you can merge with master once and let tests run again to see what might break, if anything. As you note, it's likely not something related to your changes. Stay tuned and I'll guide you along as much as possible. Once CI is passing, we'll do a review together to move forward.
